### DISCUSSION TITLE
Is Partial Hydration on the road map?

### DISCUSSION THREAD

**Discussion by AndrewLeedham** [remix-run/remix#680] [Q&A] [answered]
The ability to disable JavaScript is great, but I am curious if there is any plan to extend this to support Partial Hydration? Meaning only certain parts of the page are hydrated by React rather than all or nothing.

**Comment by sergiodxa** [comment, marked as answer]
Once React Server Component are stable I'm sure Remix will start using them internally and you will have that feature automatically.

**Reply by AndrewLeedham** [reply]
Makes sense. Took a while to get my head around how Server Components would support this, but I get it now. I guess #UseThePlatform, also applies to React 😄

**Reply by redbar0n** [reply]
For anyone who wants to dive deeper into Ryan's thinking shared in this thread, here is a link to the relevant post with his thinking on RSC: [React Server Components and Remix](https://remix.run/blog/react-server-components) (Dec 7, 2021)

**Comment by ryanflorence** [comment]
We're experimenting with RSC a lot, but we're not sure partial hydration actually helps the UX. If it does, we're all in, but right now it's still just a theory.

The main point of it is to get JS out of browser bundles. Remix already makes it easy to get a ton of code out of your browser bundles by moving it into loaders and actions. In the context of Remix, partial hydration isn't a guaranteed win.

But it might help too! We'll see.

**Reply by tigerabrodi** [reply]
Very interesting!

I spoke with friends over in the Solid.js community about partial hydration, great to hear the Remix team is observing and looking into it :pray:

**Reply by khromov** [reply]
@ryanflorence While Remix allows you to put some code in the server bundle, don't you still have to ship all of your React components twice - first in the SSR HTML and then again as their respective JS bundles? 

This is what the real problem, and specifically doing this for *non-interactive* components.

Many components on a typical blog or editorial page are not interactive and wouldn't need to be hydrated at all. This is important because large JS bundles equals bad PageSpeed scores (due to high CPU cost associated with parsing and executing JS) which leads to being penalized by Google and less traffic for your site.

**Reply by kiliman** [reply]
That "double" resource hit only applies to the initial page load. It won't download bundles for routes you haven't hit, and since those bundles are cached indefinitely, the costs are amortized across the entire user session. Unless your home page includes the kitchen sink, I don't think it would be an issue for most sites.

**Reply by khromov** [reply]
@kiliman 
> That "double" resource hit only applies to the initial page load.

Indeed, but Google calculates the PageSpeed ranking from an empty cache, so you get hit with that penalty - there is no concept of a user session. Google also collects real user data via Chrome and that will hopefully reflect the amortized user session performance but it's only one of several ranking factors.

> I don't think it would be an issue for most sites.

On a fresh install of Remix, with only the "Welcome to Remix" starter page, there is 183 KB of JavaScript served. This is already a non-trivial amount and that's even before you add any application logic.

**Reply by jmjpro** [reply]
@khromov When you describe that much javascript being delivered to the client, are you referring to the development build or the production build?

**Reply by khromov** [reply]
@jmjpro I'm referring to the production build. Here's an example so you can see it for yourself:
https://remix-hello-world-demo.vercel.app/

After opening the inspector and filtering the JS files you'll see this:
![2022-02-17 00_19_45-](https://user-images.githubusercontent.com/1207507/154373983-3f9f653c-69d2-4da8-8603-d89cacae3349.png)

You might ask what's up with the "67.2KB transferred" number. That's the amount of data sent over the wire. It's smaller because most web servers apply gzip compression on the data. But at the end of the day, your poor device still has to uncompress, parse and execute the full 194KB of JavaScript. That's the equivalent of roughly 194 000 characters, or ~130 A4 pages of text. And that's before you add any of your code.

PS. This is not strictly a Remix problem, you can get similar results from the Hello World Next.js app, although I think that one came in a bit lower on the total JS size last time I tried.

**Reply by ccssmnn** [reply]
The Hello World examples are not the best examples to show the benefits of Partial Hydration.

React works with a V-DOM (ReactDOM). React renders to its own DOM before updating the browsers DOM. When rendering on the server, you need to send the V-DOM state (JS) to the client, so React can perform hydration and take over. [Gatsby Docs about hydration](https://www.gatsbyjs.com/docs/conceptual/react-hydration/) The more components you have, the more hydration state needs to be sent.

Partial Hydration will eliminate the parts of the ReactDOM, which only include static components. You still have to send React itself and the Remix client libraries, but you can send less DOM state. Ryan Carniato from the Marko team has put together a demo [Tweet](https://twitter.com/RyanCarniato/status/1489660462367080448?s=20&t=xEFkaydKrhN0FVk86Wg1Xw), where many static components are rendered which don't require hydration.

**Reply by SantoJambit** [reply]
I'd like to argue, that partial hydration is also reducing the hydration data (JSON). I'm currently working on a news-website, where we initially had **20MB** of HTML due to all of the SSR fetched JSON being added to the HTML in order to hydrate the components in the browser. We reduced this to **1MB** by making certain components "static", which is still far from ideal, especially for SEO.

Since the proprietary framework we are using is using an opt-out strategy, this is a painful task and involves quite some hacky workarounds to accomplish this.

I'd like to see a framework, where this feels more natural/non-hacky. I haven't tried remix yet, just watched the "Introducing Remix v1" video, so maybe I'm missing the important part that helps with this sort of problem. Will give it a try soon.

Not sure if this is the same argument @ccssmnn was making, but I wanted to add a real-world example to the discussion.

**Comment by jaidetree** [comment]
I might have a use-case that would make Remix happier if partial hydration was possible.

```html
<div id="price-calculator" dangerouslySetInnerHTML={{__html: htmlSnapshotOfPriceCalc }} />
<script src="https://someother.website/widget.js" type="text/javascript" />
<script type="text/javascript" dangerouslySetInnerHTML={`SomeWidget.render(document.getElementById('price-calculator'))`} />
```

- There is a lot of SSO value in rendering the html snapshot from the server instead of an empty div with client-side only rendering
- It works, but in development a server-client code mismatch is thrown
- If I use the Remix guide example to set route.handle?.hydrate and opt out of all the client-side nav, it breaks the headless-ui JS in our navbar.
- The script replaces whatever is inside of `<div id="price-calc"></div>` so hydration is not needed here.

I'm thinking partial hydration would let the headless-ui components in the navbar via layout routes have their client-side counterparts to be functional, while preventing the page content widget area from trying to hydrate.

**Reply by sergiodxa** [reply]
Until Remix stops using React or React Server Components becomes stable there will be no way to have partial hydration out of the box.

You could do it yourself, by disabling Remix client-side JS and sending your own JS to hydrate parts of the app.

When Remix starts adding support for another UI lib that supports partial hydration, or when RSC is stable and Remix supports it, it will be possible
